crypto/init.c

```diff
@@ -729,3 +680,49 @@ int OPENSSL_atexit(void (*handler)(void))
     return 1;
 }
 
+/*
+ * Library  destructor (a.k.a. _fini()). Code here is derived
+ * from stuff found in providers/fips/self_test.c
+ *
+ * We do kind of  best effort cleanup. If given platform/compiler
+ * does not support module destructors then its up to application
+ * to explicitly call OPENSSL_cleanup();
+ */
+#if defined(__GNUC__) && !defined(_AIX)
+
+void call_cleanup(void);
+
+void __attribute__((destructor))
+call_cleanup(void)
+{
+    if (no_cleanup == 0)
+        OPENSSL_cleanup();
+}
+
+#elif defined(__sun)
+# pragma fini(OPENSSL_cleanup)
```

Comment:
Why the different handling for __sun and _AIX?

Reply:
You actually found a bug for `__sun` branch. It is the same bug I was hunting down on windows (failure in shlibloadtest). We need to check `no_clenup` flag to see if `OPENSSL_cleanup()` should be called. This is the update I'm going to submit in next PR iteration:
```
702 }                                                                                                                                                                                                                                                                                                                       
703
704 #elif (defined(__sun) || defined(_AIX)) && !defined(__GNUC__)
705 void _cleanup(void);
706 # pragma fini(_cleanup)
707 void _cleanup(void)
708 {
709     if (ossl_no_cleanup == 0)
710         OPENSSL_cleanup();
711 }
712
```